internal/operator/uninstall.go

```diff
@@ -69,82 +81,116 @@ func (u *Uninstall) Run(ctx context.Context) error {
 	}
-
-	installPlanKey := types.NamespacedName{
-		Namespace: sub.Status.InstallPlanRef.Namespace,
-		Name:      sub.Status.InstallPlanRef.Name,
+	catsrc.TypeMeta = metav1.TypeMeta{
+		Kind:       v1alpha1.CatalogSourceKind,
+		APIVersion: v1alpha1.CatalogSourceCRDAPIVersion,
 	}
 
 	// Since the install plan is owned by the subscription, we need to
 	// read all of the resource references from the install plan before
 	// deleting the subscription.
-	deleteObjs, err := u.getInstallPlanResources(ctx, installPlanKey)
-	if err != nil {
-		return err
+	var crds, csvs, others []controllerutil.Object
+	if sub.Status.InstallPlanRef != nil {
+		ipKey := types.NamespacedName{
+			Namespace: sub.Status.InstallPlanRef.Namespace,
+			Name:      sub.Status.InstallPlanRef.Name,
+		}
+		var err error
+		crds, csvs, others, err = u.getInstallPlanResources(ctx, ipKey)
+		if err != nil {
+			return fmt.Errorf("get install plan resources: %v", err)
+		}
 	}
 
 	// Delete the subscription first, so that no further installs or upgrades
 	// of the operator occur while we're cleaning up.
-	if err := u.config.Client.Delete(ctx, sub); err != nil {
-		return fmt.Errorf("delete subscription %q: %v", sub.Name, err)
-	}
-	u.config.Log("subscription %q deleted\n", sub.Name)
-
-	// Ensure CustomResourceDefinitions are deleted first, so that the operator
-	// has a chance to handle CRs that have finalizers.
-	sort.SliceStable(deleteObjs, func(i, j int) bool {
-		return deleteObjs[i].GetObjectKind().GroupVersionKind().Kind == "CustomResourceDefinition"
-	})
-	for _, obj := range deleteObjs {
-		err := u.config.Client.Delete(ctx, obj)
-		if err != nil && !apierrors.IsNotFound(err) {
+	if err := u.deleteObjects(ctx, false, sub); err != nil {
+		return err
+	}
+
+	if u.DeleteCRDs {
+
+		// Ensure CustomResourceDefinitions are deleted first, so that the operator
+		// has a chance to handle CRs that have finalizers.
+		if err := u.deleteObjects(ctx, true, crds...); err != nil {
 			return err
 		}
-		if err == nil {
-			u.config.Log("%s %q deleted\n", strings.ToLower(obj.GetObjectKind().GroupVersionKind().Kind), obj.GetName())
-		}
+	}
+
+	// Delete CSVs and all other objects created by the install plan.
+	objects := append(csvs, others...)
+	if err := u.deleteObjects(ctx, false, objects...); err != nil {
+		return err
 	}
 
 	// Delete the catalog source. This assumes that all underlying resources related
 	// to this catalog source have an owner reference to this catalog source so that
 	// they are automatically garbage-collected.
-	if err := u.config.Client.Delete(ctx, catsrc); err != nil {
-		return fmt.Errorf("delete catalog source: %v", err)
+	if err := u.deleteObjects(ctx, false, catsrc); err != nil {
+		return err
 	}
-	u.config.Log("catalogsource %q deleted\n", catsrc.Name)
 
 	// If this was the last subscription in the namespace and the operator group is
 	// the one we created, delete it
-	if len(subs.Items) == 1 {
+	if u.DeleteOperatorGroups && len(subs.Items) == 1 {
```

Comment:
We should always be cleaning up the SDK's operator group if no more subs are bound to it.

Reply:
I agree. All usages set `u.DeleteAll` to true and `u.DeleteOperatorGroupNames` to `[]string{operator.SDKOperatorGroupName}`.

I'm shooting to eventually make kubectl operator plugin's [`action.OperatorUninstall`](https://github.com/operator-framework/kubectl-operator/blob/main/internal/pkg/action/operator_uninstall.go) an exported function so that we can re-use the code instead of maintaining it in two places.